--- glibc-2.21.orig/hurd/Versions
+++ glibc-2.21/hurd/Versions
@@ -124,6 +124,14 @@ libc {
     # functions used in macros & inline functions
     __errno_location;
   }
+  GLIBC_2.21 {
+    # functions used by libpthread and <hurd/signal.h>
+    _hurd_sigstate_set_global_rcv;
+    _hurd_sigstate_lock;
+    _hurd_sigstate_pending;
+    _hurd_sigstate_unlock;
+    _hurd_sigstate_delete;
+  }
 
   HURD_CTHREADS_0.3 {
     # weak refs to libthreads functions that libc calls iff libthreads in use
