--- a/csu/Makefile
+++ b/csu/Makefile
@@ -47,6 +47,9 @@ tests-static := tst-empty
 
 ifeq (yes,$(build-shared))
 extra-objs += S$(start-installed-name) gmon-start.os
+ifneq ($(start-installed-name),$(static-start-installed-name))
+extra-objs += gmon-start.o
+endif
 install-lib += S$(start-installed-name)
 generated += start.os
 else
@@ -98,10 +101,14 @@ endif
 # startfile with gmon-start.o, which defines a constructor function
 # to turn on profiling code at startup.
 ifeq (yes,$(build-shared))
-$(addprefix $(objpfx),$(sort g$(start-installed-name) \
-			     g$(static-start-installed-name))): \
+$(objpfx)g$(start-installed-name): \
   $(objpfx)g%: $(objpfx)S% $(objpfx)gmon-start.os
 	$(link-relocatable)
+ifneq ($(start-installed-name),$(static-start-installed-name))
+$(objpfx)g$(static-start-installed-name): \
+  $(objpfx)g%: $(objpfx)% $(objpfx)gmon-start.o
+	$(link-relocatable)
+endif
 else
 $(addprefix $(objpfx),$(sort g$(start-installed-name) \
 			     g$(static-start-installed-name))): \
