--- glibc-2.22/sysdeps/mach/hurd/bits/param.h	2015-08-05 08:42:21.000000000 +0200
+++ glibc-2.22/sysdeps/mach/hurd/bits/param.h	2016-12-24 17:39:46.421987830 +0100
@@ -64,6 +64,7 @@
 
 #define	GNU	1994100		/* GNU version (year, month, and release).  */
 
+#define EXEC_PAGESIZE 4096
 
 /* BSD names for some <limits.h> values.  We do not define the BSD names
    for the values which are not statically limited, such as NOFILE.  */
